File: solutions/python/advanced-graphs/redundant-connection.py
# LeetCode 684 — Redundant Connection (Medium)
# Category: Advanced Graphs · Approach: Union-Find
# Time: O(n * α(n)) | Space: O(n)
# Source: https://leetcode.com/problems/redundant-connection/

def find_redundant_connection(edges: list[list[int]]) -> list[int]:
    n = len(edges)
    parent = list(range(n + 1))

    def find(x: int) -> int:
        while parent[x] != x:
            parent[x] = parent[parent[x]]
            x = parent[x]
        return x

    for u, v in edges:
        ru, rv = find(u), find(v)
        if ru == rv:
            return [u, v]
        parent[ru] = rv
    return []

File: solutions/typescript/advanced-graphs/redundant-connection.ts
// LeetCode 684 — Redundant Connection (Medium)
// Category: Advanced Graphs · Approach: Union-Find
// Time: O(n * α(n)) | Space: O(n)
// Source: https://leetcode.com/problems/redundant-connection/

function findRedundantConnection(edges: number[][]): number[] {
  const n = edges.length;
  const parent = Array.from({ length: n + 1 }, (_, i) => i);
  const find = (x: number): number => {
    while (parent[x] !== x) {
      parent[x] = parent[parent[x]];
      x = parent[x];
    }
    return x;
  };
  for (const [u, v] of edges) {
    const ru = find(u);
    const rv = find(v);
    if (ru === rv) return [u, v];
    parent[ru] = rv;
  }
  return [];
}
